미션 제출합니다.

seungy0 opened the pull request.

seungy0 committed:
docs: 기능 요구사항 작성

seungy0 committed:
feat: 숫자값에 따라 자동차 전진하거나 멈추는 기능 구현

0에서 3까지는 멈춤, 4에서 9까지는 전진

seungy0 committed:
feat: 랜덤값에 따라 자동차 전진하거나 멈추는 기능 구현

seungy0 committed:
docs: 진행사항 반영

seungy0 committed:
feat: 자동차 이름을 입력받는 기능 구현

- 자동차 이름을 입력받아 racingCar에 저장
- MVC 패턴 적용

seungy0 committed:
feat: 시도할 횟수 입력 구현

seungy0 committed:
feat: 각 차량의 이름과 이동거리 출력 기능 구현

trial 횟수만큼 이동하고 그 다음에 각 차량 이동거리 출력함

seungy0 committed:
feat: 우승자 판별 및 출력

seungy0 committed:
feat: 자동차 이름 유효성 검사

seungy0 committed:
chore: racingcar 패키지로 코드 이동

racingcar와 caculator를 분리하기 위함

seungy0 committed:
docs: 문자열 덧셈 계산기 요구사항 정리

seungy0 committed:
docs: 기능 요구사항 세분화

seungy0 committed:
feat: 구분자를 기준으로 문자 분할

seungy0 committed:
feat: 커스텀 구분자 지정

seungy0 committed:
feat: 입력값 유효성 검사, 숫자 합 출력

seungy0 committed:
style: 코드 스타일 적용

seungy0 committed:
refactor: 메서드명 변경

getter가 아닌데 get을 접두어로 사용하는 메서드명 수정

Rok93 commented on README.md:
```diff
@@ -1,9 +1,29 @@
+
+### 자동차 경주
+
+- [x] 랜덤값으로 전진하거나 멈추기 (0에서 3까지는 멈춤, 4에서 9까지는 전진)
```
기능 구현 목록표 잘 작성해주셨네요 👍💯

조금 더 첨언을 드리자면 세부 사항들은 아래와 같이 depth를 들여서 기능 구현 목록을 작성해주시면 가독성이 좋을 것 같아요. 😉
```suggestion
- [x] 랜덤값으로 전진하거나 멈추기
  - [x] 0 ~ 3 : 정지
  - [x] 4 ~ 9 : 전진
```

Rok93 commented on README.md:
```diff
@@ -1,9 +1,29 @@
+- [x] 자동차 이름을 입력받는다.
+- [x] 시도할 횟수를 입력받는다.
+- [x] 각 차량의 이름과 이동거리를 출력한다.
+- [x] 우승자를 출력한다.
```
기능구현 목록표를 `입력`과 `출력`의 기준으로 잡는 것 보다는 아래와 같이 객체의 역할처럼 작성하는게 기능단위로 구현하기가 조금 더 수월할 것 같아요. 😃

```suggestion
- [ ] 자동차는 전진 혹은 정지할 수 있다. 
- [ ] 우승자를 선별한다. 
  - [ ] 우승자는 한명 이상 존재할 수 있다.
- [ ] 시도 횟수만큼 레이스를 진행한다.
```

Rok93 commented on src/main/java/calculator/CalculatorApplication.java:
```diff
@@ -0,0 +1,17 @@
+        String input = scanner.nextLine();
+        System.out.println("Result: " + calculatorController.add(input));
+
+        scanner.close();
```
아래와 같이 `Scanner` 같은 `Closeable` 구현체는 [try with resource](https://www.baeldung.com/java-try-with-resources)를 활용하도록 변경해볼 수 있을 것 같아요. 😃

```suggestion
        try(Scanner scanner = new Scanner(System.in)) {
            CalculatorController calculatorController = new CalculatorController();

            System.out.println("Enter the input: ");
            String input = scanner.nextLine();
            System.out.println("Result: " + calculatorController.add(input));
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
```

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+
+public class CalculatorController {
+
+    Delimiter delimiter;
```
`CalculatorController`가 자기자신이 아닌 외부에서 이 상태 값에 바로 접근할 수 없도록 접근제한자를 `private`로 제한해주세요. 

아래의 글이 도움이되실 것 같아요. 😉
[캡슐화란-무엇인가-어떤-이점이-있는가](https://bperhaps.tistory.com/entry/%EC%BA%A1%EC%8A%90%ED%99%94%EB%9E%80-%EB%AC%B4%EC%97%87%EC%9D%B8%EA%B0%80-%EC%96%B4%EB%96%A4-%EC%9D%B4%EC%A0%90%EC%9D%B4-%EC%9E%88%EB%8A%94%EA%B0%80)

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+
+    public Optional<String> parseCustomDelimiter(String input) {
+        String regex = "//(.)";
+        Pattern pattern = Pattern.compile(regex);
```
[불필요한 객체 생성을 피하라](https://velog.io/@shinmj1207/Effective-Java-%EA%B0%9D%EC%B2%B4-%EC%83%9D%EC%84%B1%EA%B3%BC-%ED%8C%8C%EA%B4%B46) 글을 참고해서 값비싼 Pattern 객체를 상수로 추출해서 재사용해보면 어떨까요? 🙂

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+        return Optional.empty();
+    }
+
+    public void validateInput(String input) {
```
`parseCustomDelimiter`, `validateInput`, `parseNumber`은 외부에서 사용하지 않는 메서드인 것 같은데 접근제한자를 `public`으로 두신 이유가 있을까요? 🤔

외부에서 사용하지 않는 메서드라면 `private`으로 제한해주세요. 🙏

[접근 제한자 참고자료](https://hongong.hanbit.co.kr/%EC%9E%90%EB%B0%94-%EC%A0%91%EA%B7%BC-%EC%A0%9C%ED%95%9C%EC%9E%90%EC%9D%98-%EC%A2%85%EB%A5%98%EC%99%80-%EC%B0%A8%EC%9D%B4public-protected-private/)

seungy0 replied:
감사합니다.
validateInput 메서드는 Test Code에서 호출하는데 접근제한자를 private로 수정하면 테스트가 어려울 것 같습니다.
이럴 떄는 어떻게 하는게 좋을까요?

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+            .collect(Collectors.toList());
+    }
+
+    public int add(String input) {
```
다른 객체와 협력하기위한 public 메서드를 가장 상단으로 끌어올리고 클래스 내부에서만 쓰이는 메서드들을 밑으로 내려주시면 조금 더 코드 파악하는데 도움이될 것 같아요. 😉


추가적으로 내부적으로만 쓰이는 메서드들은 호출되는 순서대로 나열하면 마찬가지로 코드를 파악하는데 더 좋을 것 같습니다.

Rok93 replied:
실제로 협력을위한 메서드는 이 `add` 메서드인 것 같은데요. 
이 메서드는 test 대상으로 보이네요. 😃 
테스트 코드 작성해주시면 좋을 것 같습니다. 😉

가급적이면 한 개의 input에 대한 테스트보다는  여러 input에 대한 테스트를 작성하면 더 좋은 테스트 코드가 될 것 같네요.

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+        delimiter.addCustomDelimiter(parseCustomDelimiter(input).orElse(""));
+        List<Integer> numbers = parseNumber(input);
+        return numbers.stream()
+            .reduce(0, Integer::sum);
```
stream 활용 💯👍
아래와 같은 방식으로도 변경해볼 수 있을 것 같아요. 😃

```suggestion
        return numbers.stream()
            .mapToInt(Integer::intValue)
            .sum();
```

Rok93 commented on src/main/java/calculator/Delimiter.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class Delimiter {
```
사소할 수 있지만 내부적으로 복수개의 구분자를 가지고있으니 클래스명도 `Delimiters`와 같이 복수형으로 변경해주면 
네이밍만으로 구분자를 여러개 가졌다는 것을 쉽게 파악할 수 있을 것 같아요. 🙂

Rok93 commented on src/main/java/calculator/Delimiter.java:
```diff
@@ -0,0 +1,26 @@
+    List<String> delimiters;
+
+    public Delimiter() {
+        delimiters = new ArrayList<>(Arrays.asList(",", ":"));
```
`Arrays.asList(",", ":")`는 어떤 의미를 가질까요? 
이 코드를 처음보는 사람이라면 의미를 파악하기 어려울 것 같아요. 🥲
해당 ArrayList를 상수로 추출하고 적절한 `상수명`을 부여한다면 조금 더 의미를 명확하게 파악할 수 있을 것 같아요. 😃

아래의 글을 참고해보시면 도움이되실 것 같아요. 😉
(엄밀하게 말하면 조금 다르지만, 숫자이냐 다른 값이냐의 차이인 것 같아요)
[의미가 불분명한 매직 넘버를 상수로 선언하라](https://javabom.tistory.com/28)

Rok93 commented on src/main/java/calculator/Delimiter.java:
```diff
@@ -0,0 +1,26 @@
+        delimiters = new ArrayList<>(Arrays.asList(",", ":"));
+    }
+
+    public List<String> split(String s) {
```
메서드 파라미터 네이밍을 조금 더 의미를 명확하게 드러낼 수 있는 네이밍으로 변경해보면 어떨까요? 😃

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class CalculatorController {
```
이 클래스의 객체는 문자열을 입력받아 덧셈(= add)을 하고있어요. 
이 클래스는 실질적으로 `문자열 계산기`의 역할을 하고있는 것 같은데, 승규님은 어떻게 생각하시나요? 🤔

CalculatorController 대신에 `Calculator` 혹은 문자열 계산기 도메인에 더 fit하게 `StringCalculator` 같은 네이밍으로 변경해보시면 어떨까요? 🙌

Rok93 commented on src/main/java/calculator/CalculatorController.java:
```diff
@@ -0,0 +1,53 @@
+        Matcher matcher = pattern.matcher(input);
+
+        if (matcher.find()) {
+            return Optional.ofNullable(matcher.group(1));
```
Optional 활용 👍💯

조금 고민해볼부분은 CalculatorController가 너무 많은 역할을 하고있다고 느껴지는데 `문자열`에서 `구분자`와 `수식`을 분리하는 책임을 새로운 객체에게 위임해보면 어떨까요? 😃

새롭게 생성한 객체(A)가 문자열에서 추출한 결과물(= 구분자)은 `Delimiter` 객체가 되겠군요? 😃
또한 A가 문자열에서 추출해낸 수식을 가지고 CalculatorController 객체는 A가 추출해낸 두 결과물들과 협력하여 덧셈을 수행하면 될 것 같네요.

Rok93 replied:
이 `parseCustomDelimiter` 의 결과를 Optional로 감싼 것은 나쁘지 않은 접근이라 생각하는데, 실제로 이 메서드를 사용하는 곳들은 적절하게 사용된 느낌이 아니라 조금 아쉽게 느껴지네요. 🤔

`delimiter.addCustomDelimiter(parseCustomDelimiter(input).orElse(""));` 와 같이 `customDelimiter`가 없는 경우에는 굳이  `빈 문자열`을 Delimiter 객체에 추가해줄 필요가 없는 것 같은데, 컬렉션에 불필요한 요소를 추가해주고있는 것 같아요. 🤔

아래와 같이 Optional의 `ifPresent` 함수를 활용하여 cusomDelimiter가 존재하지 않는 경우에는 컬렉션에 요소를 추가해주지 않는 것이 좋은 것 같아요. 👀

```java
parseCustomDelimiter(input).ifPresent(delimiter::addCustomDelimiter);
```

아래의 글을 참고하시면 Optional을 사용하시는데 도움이되실 것 같습니다. 🙌
[Java Optional 바르게 쓰기](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+    private int moves = 0;
+
+    public RacingCar() {
+    }
```
기본 생성자를 정의하신 이유가 있을까요? 🤔
기본생성자를 굳이 사용하는 상황이 아니라면 제거해주세요.

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public boolean moveOrNot(int num) {
+        return num >= 4 && num <= 9;
```
`4`, `9`와 같은 수는 매직넘버로 보이네요. 😃
위에서 남긴 코멘트와 동일하게 상수로 추출해주세요. 🙏

사소한 팁이지만  조건식을  `return 4 <= num && num <= 9;` 와 같이 두면 조금 더 조건식이 잘 읽힐 것 같아요. 😉

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+public class RacingCar {
+
+    private String name;
+    private int moves = 0;
```
[구글 자바 코드 컨벤션](https://google.github.io/styleguide/javaguide.html#s5.2.5-non-constant-field-names)에 따르면 필드 이름은 `명사` 혹은 `명사구`를 사용하라고 가이드하고있어요. 

컨벤션에 맞게 다른 적절한 네이밍으로 변경해보면 어떨까요? 😃

추가적으로 변수명과 관련된 아래의 글도 같이 읽어보시면 도움이되실 것 같아요. 😉
[좋은 코드를 위한 자바 변수명 네이밍](https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming/)

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    public void move() {
```
getter를 제외한 모든 public 메서드에 대해서 테스트를 작성해주세요. 🙏

seungy0 replied:
감사합니다.
<img width="479" alt="스크린샷 2024-03-27 오후 6 03 16" src="https://github.com/next-step/java-racingcar-kakao/assets/51262434/0aa0830c-c9ba-4d9b-bc10-0fd1f30e0e32">

랜덤값을 사용하는 메서드를 제외하고 모든 메서드를 테스트하고 있습니다. 혹시 제가 빠트린 부분이 있을까요?ㅎㅎ

Rok93 replied:
제가 설명이 조금 부족했네요! 😅
제가 말씀드리고 싶었던 이야기는 이런 랜덤값을 사용하는 메서드도 테스트 코드를 작성해달라는 이야기였어요! 😃

Rok93 replied:
제가 설명이 조금 부족했네요! 😅
제가 말씀드리고 싶었던 이야기는 이런 랜덤값을 사용하는 메서드도 테스트 코드를 작성해달라는 이야기였어요! 😃

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+        this.name = name;
+    }
+
+    public void getNumAndMove(int num) {
```
RacingCar 클래스 내부에서만 쓰이는 메서드이니 접근제한자를 `private`로 제한해주세요. 🙏

seungy0 replied:
감사합니다.
이 부분도 테스트 코드에서의 호출을 위해 public을 사용했는데요ㅎㅎ
찾아보니 private 메소드도 Java Reflection을 사용해 테스트가 가능하긴 하네요.
이런식으로 고치는게 좋을까요?

Rok93 replied:
물론 Reflection을 사용하면 private 메서드를 테스트할 수 있지만 이는 권장되는 방식이아닙니다. 😃

테스트의 대상은 다른 객체들과 협력을 위해 사용되는(= public으로 제공되는, getter/setter 등은 제외) 메서드들을 테스트해야하는데, 
그렇다면 현재 테스트 대상은 `move` 메서드로 보이네요. 
하지만 이 메서드에대한 테스트를 작성하시는 대신 이 메서드가 내부적으로 호출하고있는 `moveIfNumBig` 메서드에 대한 테스트만 작성하셨네요. 

아마도 move 함수가 테스트하기 어렵다고판단하셔서 지금처럼 차선책으로 내부에서 호출하는 메서드를 호출하신것으로 보입니다. 😃
저희는 이렇게 테스트하기 어려운 메서드를 테스트할 수 있는 코드들로 변경하여 테스트 영역을 더 확보해서 최대한 테스트 코드로 비즈니스 로직을 보호해야하는데요. 

일단 
Car 클래스 안에있는 `AS IS` 코드들을 `TO BE` 코드로 변경해보고 
### AS IS
```
    public void move() {
        int num = makeRandomNum();
        moveIfNumBig(num);
    }

    public void moveIfNumBig(int num) {
        if (moveOrNot(num)) {
            this.position++;
        }
    }

    public boolean moveOrNot(int num) {
        return MOVE_MIN_LIMIT <= num && num <= MOVE_MAX_LIMIT;
    }
```

사실 위의코드는 아래의 메서드 하나로도 이해가 가능한 것 같습니다. 😃

### TO BE 
```
    public void move() {
        int num = makeRandomNum();
        if (MOVE_MIN_LIMIT <= num && num <= MOVE_MAX_LIMIT) {
            this.position++;
        }
    }
```

`TO BE`의 move 메서드는 왜 테스트하기 힘들까요? 
테스트하기 어려운 요소가 존재하기 때문인데요. 2단계 미션 소개 페이지를 참고하시면 아래와 같은 가이드를 주고있어요. 

> 테스트 가능한 부분과 테스트하기 힘든 부분을 분리해 테스트 가능한 부분에 대해서만 단위 테스트를 진행한다.

다시 코드로 돌아와서 move 함수 내부에서 어떤 요소가 가장 테스트하기 어려운 요소로 보이시나요? 😃
이러한 테스트하기 어려운 요소를 최대한 외부로 뽑아낸다면 테스트하기가 조금 더 쉬워지지 않을까요? 

잘 이해가되지 않으신다면 랜덤 값으로 받던 요소를 move 메서드의 메서드 파라미터로 주입받도록 변경해보시고 잘 이해가되지 않는다면 아래의 영상을 참고해보시면 좋을 것 같습니다. 😃

[무엇을 테스트할 것인가? 어떻게 테스트할 것인가? (권용근)](https://www.youtube.com/watch?v=YdtknE_yPk4)

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private static int makeRandomNum() {
+        Random random = new Random();
```
`makeRandomNum` 메서드를 호출할 때, Random 객체를 매번 생성하게될 것 같아요. 🤔
Random 객체를 상수로 추출하고 재사용하도록 변경해보면 어떨까요? 😃

seungy0 replied:
감사합니다.
Random 객체는 한 메소드에서만 사용해서 필드에 상수로 추출하기 조금 애매하다고 느껴졌습니다.

필드에 상수로 추출할 시 메모리에 공간을 계속 차지할 것 같아서 장단점이 있을 것 같습니다ㅎㅎ
위에서 링크 주신 [불필요한 객체 생성을 피하라](https://velog.io/@shinmj1207/Effective-Java-%EA%B0%9D%EC%B2%B4-%EC%83%9D%EC%84%B1%EA%B3%BC-%ED%8C%8C%EA%B4%B46) 의 마지막에 JVM 최적화 부분을 보고 Random 객체는 이대로 남겨두어도 되는건가? 하는 생각이 들었습니다.
제가 놓친 부분이 있다면 알려주세요 :)

Rok93 replied:
`makeRandomNum`는 아마도  `시도횟수 x 자동차 댓수` 만큼 호출될 것으로 생각됩니다. 
지금의 코드의 경우 `Random` 객체를 호출된 갯수만큼 생성할거에요. 

시도횟수 x 자동차 댓수 만큼 수행하더라도 자동차 경주 게임은 매우 작은 서비스이기 때문에 문제가 생기지 않겠지만 
만약 시도횟수 x 자동차 댓수가 늘어난다면 엄청난 양의 객체를 매번 생성하게될거에요. 
불필요하게 객체를 많이 생성하는 것 보다는 이런 요소는 상수로뽑아놓고 재사용을하는 것이 조금 더 자원을 절약할 수 있는 방법이기 때문에 말씀드렸어요. 😃

객체 하나를 메모리 상에 적재해두는 것과 매번 객체를 생성해서 힙 영역에 저장되고 GC가 동작으로 제거되는 행위를 반복하는 것 중에서 어떤 점이 효율적일까요? 🤔

저는 전자가 더 좋다고 생각하는데 승연님은 어떻게 생각하시나요?

Rok93 reviewed (changes requested):
승규님 안녕하세요.
자동차 경주 게임을 함께하게된 김경록입니다. 🙇‍♂️

1단계 구현 잘해주셨습니다. 👍
몇몇 코멘트 남겨두었으니 확인해서 반영해주세요. 🙏
미션 진행하시면서 어려운 부분 혹은 이해가 잘 안되는 부분이있으면 언제든지 DM 주세요. 🙂

seungy0 committed:
docs: 기능 요구사항 가독성 개선, 객체 역할별 분류

기능 요구사항 depth 분리, 객체별 요구사항 분류

seungy0 committed:
refactor: calculator 리팩토링

- Scanner의 close-with-resources 사용
참고: https://robin00q.tistory.com/76

- private 접근제한자 지정

- 합 구하는 방식 reduce -> mapToInt로 변경
성능이 for-loop > mapToInt > reduce 순으로 좋음.
참고: https://bcp0109.tistory.com/299

- Delimiter -> Delimiters 클래스명 변경

- 상수 지정
",", ";"와 같은 하드코딩된 요구사항 상수로 지정

- 의미없는 메서드 파라미터명 변경
String s 와같은 파라미터명 수정

- CalculatorController -> Calculator 로 클래스명 변경

- Calculator 클래스 add 메서드 테스트 코드 작성

- Optional 반환값에 대한 orElse -> ifPresent로 변경
불필요하게 빈 문자열을 사용하는 부분이 있었음.

- InputProcessor 클래스 작성
리뷰어분께서 말씀하신 방식과 조금 다르긴하지만 책임분리

seungy0 commented:
변경할 거리가 별로 없을거라 생각해서 커밋을 뭉뚱그려서 했는데 여러개로 쪼갤 걸 그랬네요😅

seungy0 committed:
refactor: 사용하지 않는 생성자 제거

seungy0 committed:
refactor: 생성자 제거에 따른 테스트 코드 수정

seungy0 committed:
refactor: 매직넘버 상수로 추출

seungy0 committed:
refactor: 필드명 명사로 변경

moves를 position으로 변경

seungy0 committed:
refactor: 메소드명 변경

getter와 혼동될 수 있는 메소드명 변경

seungy0 committed:
refactor: 메소드 순서 변경

private 메소드를 하단으로 내림

Rok93 commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    public void move() {
```
[이 질문](https://github.com/next-step/java-racingcar-kakao/pull/18#discussion_r1540718398)에 대한 답변은 여기에 남겨두겠습니다! 

물론 Reflection을 사용하면 private 메서드를 테스트할 수 있지만 이는 권장되는 방식이아닙니다. 😃

테스트의 대상은 다른 객체들과 협력을 위해 사용되는(= public으로 제공되는, getter/setter 등은 제외) 메서드들을 테스트해야하는데, 
그렇다면 현재 테스트 대상은 `move` 메서드로 보이네요. 
하지만 이 메서드에대한 테스트를 작성하시는 대신 이 메서드가 내부적으로 호출하고있는 `moveIfNumBig` 메서드에 대한 테스트만 작성하셨네요. 

아마도 move 함수가 테스트하기 어렵다고판단하셔서 지금처럼 차선책으로 내부에서 호출하는 메서드를 호출하신것으로 보입니다. 😃
저희는 이렇게 테스트하기 어려운 메서드를 테스트할 수 있는 코드들로 변경하여 테스트 영역을 더 확보해서 최대한 테스트 코드로 비즈니스 로직을 보호해야하는데요. 

일단 
Car 클래스 안에있는 `AS IS` 코드들을 `TO BE` 코드로 변경해보고 
### AS IS
```
    public void move() {
        int num = makeRandomNum();
        moveIfNumBig(num);
    }

    public void moveIfNumBig(int num) {
        if (moveOrNot(num)) {
            this.position++;
        }
    }

    public boolean moveOrNot(int num) {
        return MOVE_MIN_LIMIT <= num && num <= MOVE_MAX_LIMIT;
    }
```

사실 위의코드는 아래의 메서드 하나로도 이해가 가능한 것 같습니다. 😃

### TO BE 
```
    public void move() {
        int num = makeRandomNum();
        if (MOVE_MIN_LIMIT <= num && num <= MOVE_MAX_LIMIT) {
            this.position++;
        }
    }
```

`TO BE`의 move 메서드는 왜 테스트하기 힘들까요? 
테스트하기 어려운 요소가 존재하기 때문인데요. 2단계 미션 소개 페이지를 참고하시면 아래와 같은 가이드를 주고있어요. 

> 테스트 가능한 부분과 테스트하기 힘든 부분을 분리해 테스트 가능한 부분에 대해서만 단위 테스트를 진행한다.

다시 코드로 돌아와서 move 함수 내부에서 어떤 요소가 가장 테스트하기 어려운 요소로 보이시나요? 😃
이러한 테스트하기 어려운 요소를 최대한 외부로 뽑아낸다면 테스트하기가 조금 더 쉬워지지 않을까요? 

잘 이해가되지 않으신다면 랜덤 값으로 받던 요소를 move 메서드의 메서드 파라미터로 주입받도록 변경해보시고 잘 이해가되지 않는다면 아래의 영상을 참고해보시면 좋을 것 같습니다. 😃

[무엇을 테스트할 것인가? 어떻게 테스트할 것인가? (권용근)](https://www.youtube.com/watch?v=YdtknE_yPk4)

seungy0 replied:
피드백 감사합니다.
테스트 용이성을 위해 테스트가 어려운 부분을 바운더리 레이어로 옮긴다는 아이디어는 잘 알겠습니다.

다만 그럴 경우 랜덤한 값을 생성하는 부분이 Car 객체와 멀어져서 응집성이 떨어진다는 생각이 듭니다.
테스트를 고려하지 않았다면 저는 Random 값 생성하는 부분을 Car안에 두거나, 이 부분을 분리한다고 해도 Car에서 랜덤값 생성하는 코드(ex 함수형 인터페이스) 를 호출하는 방식으로 코드를 작성했을 것 같습니다.

이러한 관점에서 테스트를 위해 랜덤값 생성 부분을 바운더리 레이어로 옮기는게 꺼려지는데 어떻게 생각하시는지 궁금합니다ㅎㅎ

Rok93 replied:
> 랜덤한 값을 생성하는 부분이 Car 객체와 멀어져서 응집성이 떨어진다는 생각이 듭니다.

랜덤한 값을 생성하는 부분이 Car 객체의 역할일지 먼저 고민해보면 좋을 것 같아요. 🤔
자동차 객체의 입장에서는 전진(혹은 정지)하는 것이 중요하지 숫자는 랜덤이고 랜덤한 숫자가 몇이상이면 전진한다와 같은 전진(혹은 정지)에 대한 결정 기준까지 알 필요가 있을까 싶어요. 

이러한 전진(혹은 정지)에 대한 결정을 다른 객체에게 위임을 해주고 Car 객체는 전진 할지 정지할지 결정해주는 객체와 협력하여 전진(or 정지)하게 된다면 객체들은 각자 하나의 책임을 가지고 그만큼 응집도가 높아질 것 같아요. 🙂

응집성이 높아진다라는 생각보다는 오히려 불필요한 의존성이 강해진다고 생각되네요. 🤔

Rok93 commented on src/main/java/racingcar/model/RacingCars.java:
```diff
@@ -0,0 +1,30 @@
+
+public class RacingCars {
+
+    List<RacingCar> racingCars;
```
마찬가지로 캡슐화를 위해서 RacingCars의 인스턴스 변수의 접근제한자를 `private`로 변경해주세요! 🙏
또한 이 인스턴스 변수는 한번 할당되면 변경되지 않는 것 같으니 `final` 키워드도 추가해주시면 좋을 것 같네요. 😉

seungy0 replied:
변경했습니다.

Rok93 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,48 @@
+
+public class RacingCarController {
+
+    RacingCars racingCars;
```
Controller가 `racingCars` 인스턴스 변수를 가지는 것은 적절하지 않은 방식입니다. 

2단계에서는 `MVC 패턴` 기반으로 리팩토링을 진행하시게 될텐데 아래의 MVC 관련 참고자료들을 참고해보셔서 `racingCars`를 인스턴스 변수로 두는 것이 아니라 입력받은 `playGame` 메서드 안에서의 지역변수로 둬보시면 좋을 것 같습니다. 

[MVC 패턴 소개](https://www.youtube.com/results?search_query=mvc+%ED%8C%A8%ED%84%B4)
[웹 MVC 각 컴포넌트 역할](https://tecoble.techcourse.co.kr/post/2021-04-26-mvc/)

Rok93 commented on src/main/java/racingcar/view/RacingCarView.java:
```diff
@@ -0,0 +1,55 @@
+        System.out.println(winners + "가 최종 우승했습니다.");
+    }
+
+    private static void checkCarName(String carName) {
+        if (carName.length() > 5) {
+            throw new InValidInputException("[ERROR] 이름은 5자 이하만 가능하다.");
+        }
+    }
```
자동차 객체를 생성할 때, 매번 이 유효성 검증이 필요할 것 같아요. 🤔
이런 케이스는 RacingCar 객체를 생성하는 시점에 검증하는 것이 좋지않을까요? 😃

이 메서드를 RacingCar 클래스의 생성자로 넣어보면 어떨까요? 😉

seungy0 replied:
네 변경하겠습니다😀

Rok93 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final Delimiters delimiters;
+
+    public InputProcessor inputProcessor;
```
인스턴스 변수는 접근제한자를 private으로 둬서 외부에서 직접 접근하지 못하도록 막아주세요. 🙏
(코드 전반적으로 확인해서 누락된 부분은 반영 부탁드려요!   public 혹은 아예 없는(=default) 경우 모두 반영 대상입니다.)

seungy0 replied:
변경했습니다.

Rok93 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,33 @@
+    public InputProcessor inputProcessor;
+
+    public Calculator() {
+        this.delimiters = new Delimiters();
+        this.inputProcessor = new InputProcessor();
```
생성자 파라미터를 통해서 `Delimiters`, `InputProcessor` 객체를 주입받아서 인스턴스 변수를 초기화해주도록 변경해보면 어떨까요? 😃

seungy0 replied:
네 좋을 것 같습니다ㅎㅎ
다만 이번 미션의 경우 생성자 내부에서 생성해도 문제가 없다고 판단했습니다.

Rok93 replied:
생성자 내부에서 직접 객체를 생성하여 인스턴스 변수를 할당해주게되면 컴파일 시점에 의존성이 지정되게되어 유연성이 떨어지는 반면에 생성자 파라미터로 주입받게되면 런타임 시점에 의존성이 정해지기 때문에 더 유연하게 설계할 수 있어요. 👀

물론 지금의 경우에는 `delimiters`나 `inputProcessor`를 상속하는 구현체가 없기 때문에 이런 이점을 부각하기는 힘들겠지만 이렇게 의존성을 파라미터로 주입받게되면 테스트 코드를 작성하는 것도 더 수월해지는 이점이 있어요. 🙂

Rok93 commented on src/main/java/calculator/InputProcessor.java:
```diff
@@ -0,0 +1,29 @@
+
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n");
+
+    public void validateInput(String input) {
```
InputProcessor는 상태를 가지지 않는 Util 클래스로 보이는데, 객체 생성 없이 static 메서드로 `validateInput`, `parseCustomDelimiter`를 두는 것은 어떨까요? 🙂
이런 경우에는 기본 생성자를 `private`로 제한하면 객체 생성을 방지할 수 있어요! 

```java
    private InputProcessor() {
    }
```

seungy0 replied:
감사합니다.
현재 스펙에서 static으로 바꿔서 얻는 이점이 없다고 판단했습니다. 그리고 도메인 종속적인 유효성 검사라는 점도 static 메서드로 바꾸는 걸 주저하게 하는 것 같습니다.
static 메서드로 만드는 것도 좋은 방법이라고 생각합니다ㅎㅎ

Rok93 replied:
자신만의 고유한 상태와 행위를 가지는 것을 객체라고하는데 
`InputProcessor`는 고유한 상태 값을 가지지 않는데, **객체로써 존재할 이유**가있을까요? 
저는 오히려 이런 케이스를 매번 객체로 생성하는 비용이 조금 더 크다고 생각되는 것 같아요. 😃

Rok93 commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,48 @@
+
+    public void playGame() {
+        int trial = racingCarView.requestTrial();
+        List<RacingCar> participants = racingCars.getRacingCars();
+
+        racingCarView.startGameRound();
+
+        for (int i = 0; i < trial; i++) {
+            playGameRound(participants);
+        }
```
`racingCars` 객체는 이미 상태로 **participants**를 가지고있는데, 아래와 같이 racingCars 객체에게 메시지를 보내보면 어떨까요? 
```suggestion
        racingCarView.startGameRound();

        for (int i = 0; i < trial; i++) {
            racingCars.playGameRound();
        }
```

seungy0 replied:
view에 participants를 전달하기 위해 getter를 사용하고 있습니다.
getter를 제거할 순 없지만 RacingCars 클래스를 RacingGame으로 변경하고 내부에 playGame 메서드를 만들어서 for 문을 포한해서 해당 메서드에 넣도록하겠습니다.

Rok93 commented on src/main/java/racingcar/model/RacingCars.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public List<RacingCar> findWinners() {
+        int maxMoves = racingCars.stream()
+            .mapToInt(RacingCar::getPosition)
+            .max()
+            .orElse(0);
+
+        return racingCars.stream()
+            .filter(car -> car.getPosition() == maxMoves)
```
[getter를 사용하는 대신 객체에 메시지를 보내](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)보시면 어떨까요? 😉

seungy0 replied:
감사합니다.
getter 사용을 지양한다는 점은 이해했습니다. 다만 저는 이 경우 getter 사용이 적절하다고 생각했습니다.
만약 getter를 제거한다면 어떻게 바꿀 수 있을까요?

Rok93 replied:
아래와 같이 변경해볼 수 있을 것 같아요. 
```java
public boolean isSamePosition(int position) {
    return this.position == position;
}
```

이런 것들이 별거아닌 것 처럼 보일 수 있지만 지금같이 getter로 객체의 상태를 외부에서 꺼내서 직접 수정하게되면 (= 캡슐화를 무시) 추후 car 객체의 변경이 외부로 전파되게되어 자그마한 변경이 많은 변경을 유발하게됩니다. 
결국 이런 것 때문에 변경이 어려워지고 유지보수가 어려워지게됩니다.

Rok93 commented on src/main/java/racingcar/model/RacingCars.java:
```diff
@@ -0,0 +1,30 @@
+            .collect(Collectors.toList());
+    }
+
+    public List<RacingCar> getRacingCars() {
+        return racingCars;
+    }
```
getter, setter와 같은 메서드들은 클래스의 가장 하단으로 내려주시고, 다른 객체들과 협력하는 public 메서드들을 상위에 위치시켜주면 조금 더 코드를 파악하기 쉬울 것 같아요. 😉

seungy0 replied:
변경했습니다.

Rok93 reviewed (changes requested):
승규님 피드백 잘 반영해주셨습니다. 👍👍
조금 더 이야기해보면 좋을 부분들이 보여서 추가로 코멘트 더 남겨두었습니다. 🙏
피드백은 코멘트를 남긴위치 뿐만 아니라 코드 전체적으로 확인해서 같이 반영해주셨으면합니다. 🥲

또한 남겨주신 질문에 제 생각을 남겨두었으니 확인해주시면 될 것 같습니다. 😃

seungy0 committed:
refactor: 접근제어자 변경

public -> private final

seungy0 committed:
refactor: 메서드 순서 변경

getter 하단으로 순서 변경

seungy0 committed:
refactor: 인스턴스 변수 -> 로컬변수로 변경

seungy0 committed:
refactor: RacingCars -> RacingGame으로 책임 변경

seungy0 committed:
refactor: 필요없는 코드 정리

Rok93 reviewed (approved):
승규님 피드백 잘 반영해주셨습니다. 🙂
남겨주신 의견이나 질문들에 대해서 제 생각도 추가로 남겨두었으니
확인해주시면 좋을 것 같아요. 🙏

아쉽지만 자동차 경주 미션은 이만 종료하도록하겠습니다. 🥲
남은 미션도 화이팅입니다. 💪💪
